File: backend/compose-engine/app/services/publishers/instagram.py
"""
Instagram Graph API Reels Publishing Service
https://developers.facebook.com/docs/instagram-api/guides/content-publishing

Instagram Reels Publishing Requirements:
- Business or Creator Instagram account
- Connected Facebook Page
- 2-step process: Create container → Publish
- Video: 3-90 seconds, recommended ≤60 seconds
- Aspect ratio: 9:16 (vertical)
- Max file size: 1GB
"""

import httpx
import asyncio
from typing import Optional, Dict, Any, Literal
from dataclasses import dataclass, field
import logging

logger = logging.getLogger(__name__)

INSTAGRAM_GRAPH_BASE = "https://graph.instagram.com"
FACEBOOK_GRAPH_BASE = "https://graph.facebook.com/v21.0"


@dataclass
class InstagramReelSettings:
    share_to_feed: bool = True
    cover_url: Optional[str] = None
    thumb_offset: Optional[int] = None  # Cover frame offset in ms
    location_id: Optional[str] = None
    collaborator_usernames: list[str] = field(default_factory=list)


@dataclass
class InstagramPublishResult:
    success: bool
    media_id: Optional[str] = None
    permalink: Optional[str] = None
    error: Optional[str] = None
    error_code: Optional[int] = None


class InstagramPublisher:
    """Instagram Reels Publishing Service"""

    def __init__(self, app_id: str, app_secret: str):
        self.app_id = app_id
        self.app_secret = app_secret

    async def refresh_token(self, access_token: str) -> Dict[str, Any]:
        """
        Refresh Instagram long-lived token (extends by 60 days)
        Can only be done once per 24 hours
        """
        async with httpx.AsyncClient() as client:
            response = await client.get(
                f"{INSTAGRAM_GRAPH_BASE}/refresh_access_token",
                params={
                    "grant_type": "ig_refresh_token",
                    "access_token": access_token,
                },
            )
            result = response.json()

            if result.get("error"):
                return {
                    "success": False,
                    "error": result.get("error", {}).get("message", "Token refresh failed"),
                }

            return {
                "success": True,
                "access_token": result.get("access_token"),
                "expires_in": result.get("expires_in"),
            }

    async def get_instagram_account_from_page(
        self, access_token: str
    ) -> Dict[str, Any]:
        """Get Instagram Business Account ID from Facebook Page"""
        try:
            async with httpx.AsyncClient() as client:
                # First get user's pages
                pages_response = await client.get(
                    f"{FACEBOOK_GRAPH_BASE}/me/accounts",
                    params={"access_token": access_token},
                )
                pages_data = pages_response.json()

                if pages_data.get("error"):
                    return {
                        "success": False,
                        "error": pages_data["error"].get("message"),
                    }

                if not pages_data.get("data"):
                    return {
                        "success": False,
                        "error": "No Facebook Pages found. Instagram Business accounts require a connected Facebook Page.",
                    }

                # Get Instagram Business Account from first page with Instagram connected
                for page in pages_data["data"]:
                    ig_response = await client.get(
                        f"{FACEBOOK_GRAPH_BASE}/{page['id']}",
                        params={
                            "fields": "instagram_business_account",
                            "access_token": access_token,
                        },
                    )
                    ig_data = ig_response.json()

                    if ig_data.get("instagram_business_account"):
                        return {
                            "success": True,
                            "instagram_account_id": ig_data["instagram_business_account"]["id"],
                            "page_id": page["id"],
                        }

                return {
                    "success": False,
                    "error": "No Instagram Business Account found connected to your Facebook Pages.",
                }
        except Exception as e:
            logger.error(f"[Instagram] get_instagram_account_from_page error: {e}")
            return {"success": False, "error": str(e)}

    async def get_account_info(
        self, access_token: str, instagram_account_id: str
    ) -> Dict[str, Any]:
        """Get Instagram account info"""
        try:
            async with httpx.AsyncClient() as client:
                fields = "id,username,name,account_type,profile_picture_url,followers_count,media_count"
                response = await client.get(
                    f"{INSTAGRAM_GRAPH_BASE}/{instagram_account_id}",
                    params={"fields": fields, "access_token": access_token},
                )
                data = response.json()

                if data.get("error"):
                    return {"success": False, "error": data["error"].get("message")}

                return {
                    "success": True,
                    "id": data.get("id"),
                    "username": data.get("username"),
                    "name": data.get("name"),
                    "account_type": data.get("account_type"),
                    "profile_picture_url": data.get("profile_picture_url"),
                    "followers_count": data.get("followers_count"),
                    "media_count": data.get("media_count"),
                }
        except Exception as e:
            logger.error(f"[Instagram] get_account_info error: {e}")
            return {"success": False, "error": str(e)}

    async def _create_reels_container(
        self,
        access_token: str,
        instagram_account_id: str,
        video_url: str,
        caption: str,
        settings: Optional[InstagramReelSettings] = None,
    ) -> Dict[str, Any]:
        """
        Step 1: Create a media container for Reels
        Returns a container ID that needs to be polled for status
        """
        try:
            logger.info("[Instagram] Creating Reels container...")

            params: Dict[str, str] = {
                "media_type": "REELS",
                "video_url": video_url,
                "caption": caption[:2200],  # Instagram caption limit
                "access_token": access_token,
            }

            # Optional settings
            if settings:
                if settings.share_to_feed is not None:
                    params["share_to_feed"] = str(settings.share_to_feed).lower()
                if settings.cover_url:
                    params["cover_url"] = settings.cover_url
                if settings.thumb_offset is not None:
                    params["thumb_offset"] = str(settings.thumb_offset)
                if settings.location_id:
                    params["location_id"] = settings.location_id
                if settings.collaborator_usernames:
                    params["collaborators"] = ",".join(settings.collaborator_usernames)

            async with httpx.AsyncClient() as client:
                response = await client.post(
                    f"{INSTAGRAM_GRAPH_BASE}/{instagram_account_id}/media",
                    params=params,
                )
                data = response.json()

                if data.get("error"):
                    logger.error(f"[Instagram] Create container error: {data['error']}")
                    return {
                        "success": False,
                        "error": data["error"].get("message"),
                        "error_code": data["error"].get("code"),
                        "error_subcode": data["error"].get("error_subcode"),
                    }

                logger.info(f"[Instagram] Container created: {data.get('id')}")
                return {"success": True, "container_id": data.get("id")}

        except Exception as e:
            logger.error(f"[Instagram] _create_reels_container error: {e}")
            return {"success": False, "error": str(e)}

    async def _check_container_status(
        self, access_token: str, container_id: str
    ) -> Dict[str, Any]:
        """Check container processing status"""
        try:
            async with httpx.AsyncClient() as client:
                response = await client.get(
                    f"{INSTAGRAM_GRAPH_BASE}/{container_id}",
                    params={
                        "fields": "status,status_code",
                        "access_token": access_token,
                    },
                )
                data = response.json()

                if data.get("error"):
                    return {"success": False, "error": data["error"].get("message")}

                return {
                    "success": True,
                    "status": data.get("status"),
                    "status_code": data.get("status_code"),
                }
        except Exception as e:
            logger.error(f"[Instagram] _check_container_status error: {e}")
            return {"success": False, "error": str(e)}

    async def _wait_for_container_ready(
        self,
        access_token: str,
        container_id: str,
        max_attempts: int = 30,
        interval_sec: float = 10.0,
    ) -> Dict[str, Any]:
        """Wait for container to be ready for publishing"""
        logger.info("[Instagram] Waiting for container to be ready...")

        for attempt in range(max_attempts):
            result = await self._check_container_status(access_token, container_id)

            if not result["success"]:
                return {"success": False, "ready": False, "error": result.get("error")}

            status = result.get("status")
            logger.info(
                f"[Instagram] Container status (attempt {attempt + 1}/{max_attempts}): {status}"
            )

            if status == "FINISHED":
                return {"success": True, "ready": True}

            if status in ("ERROR", "EXPIRED"):
                return {
                    "success": False,
                    "ready": False,
                    "error": f"Container status: {status} - {result.get('status_code')}",
                }

            # Wait before next check
            await asyncio.sleep(interval_sec)

        return {
            "success": False,
            "ready": False,
            "error": "Container processing timed out",
        }

    async def _publish_container(
        self, access_token: str, instagram_account_id: str, container_id: str
    ) -> Dict[str, Any]:
        """Step 2: Publish the media container"""
        try:
            logger.info("[Instagram] Publishing container...")

            async with httpx.AsyncClient() as client:
                response = await client.post(
                    f"{INSTAGRAM_GRAPH_BASE}/{instagram_account_id}/media_publish",
                    params={
                        "creation_id": container_id,
                        "access_token": access_token,
                    },
                )
                data = response.json()

                if data.get("error"):
                    logger.error(f"[Instagram] Publish error: {data['error']}")
                    return {
                        "success": False,
                        "error": data["error"].get("message"),
                        "error_code": data["error"].get("code"),
                    }

                media_id = data.get("id")
                logger.info(f"[Instagram] Published successfully: {media_id}")

                # Get permalink
                media_response = await client.get(
                    f"{INSTAGRAM_GRAPH_BASE}/{media_id}",
                    params={"fields": "permalink", "access_token": access_token},
                )
                media_data = media_response.json()

                return {
                    "success": True,
                    "media_id": media_id,
                    "permalink": media_data.get("permalink"),
                }

        except Exception as e:
            logger.error(f"[Instagram] _publish_container error: {e}")
            return {"success": False, "error": str(e)}

    async def publish(
        self,
        access_token: str,
        instagram_account_id: str,
        video_url: str,
        caption: str = "",
        hashtags: list[str] = None,
        settings: Optional[InstagramReelSettings] = None,
    ) -> InstagramPublishResult:
        """
        Full Reels publishing flow
        Creates container → Waits for ready → Publishes

        Args:
            access_token: Instagram/Facebook access token
            instagram_account_id: Instagram Business Account ID
            video_url: URL of the video to upload
            caption: Reel caption
            hashtags: List of hashtags
            settings: Instagram Reel settings
        """
        try:
            hashtags = hashtags or []

            # Combine caption with hashtags
            if hashtags:
                hashtags_str = " ".join(
                    h if h.startswith("#") else f"#{h}" for h in hashtags
                )
                full_caption = f"{caption}\n\n{hashtags_str}"
            else:
                full_caption = caption

            # Step 1: Create container
            container_result = await self._create_reels_container(
                access_token,
                instagram_account_id,
                video_url,
                full_caption,
                settings,
            )

            if not container_result["success"]:
                return InstagramPublishResult(
                    success=False,
                    error=container_result.get("error"),
                    error_code=container_result.get("error_code"),
                )

            container_id = container_result["container_id"]

            # Step 2: Wait for container to be ready
            ready_result = await self._wait_for_container_ready(
                access_token, container_id
            )

            if not ready_result["success"] or not ready_result.get("ready"):
                return InstagramPublishResult(
                    success=False,
                    error=ready_result.get("error", "Container not ready for publishing"),
                )

            # Step 3: Publish
            publish_result = await self._publish_container(
                access_token, instagram_account_id, container_id
            )

            if not publish_result["success"]:
                return InstagramPublishResult(
                    success=False,
                    error=publish_result.get("error"),
                    error_code=publish_result.get("error_code"),
                )

            return InstagramPublishResult(
                success=True,
                media_id=publish_result.get("media_id"),
                permalink=publish_result.get("permalink"),
            )

        except Exception as e:
            logger.error(f"[Instagram] publish error: {e}")
            return InstagramPublishResult(success=False, error=str(e))

    async def get_publishing_limit(
        self, access_token: str, instagram_account_id: str
    ) -> Dict[str, Any]:
        """Get publishing quota/limits"""
        try:
            async with httpx.AsyncClient() as client:
                response = await client.get(
                    f"{INSTAGRAM_GRAPH_BASE}/{instagram_account_id}/content_publishing_limit",
                    params={
                        "fields": "quota_usage,config",
                        "access_token": access_token,
                    },
                )
                data = response.json()

                if data.get("error"):
                    return {"success": False, "error": data["error"].get("message")}

                quota_data = data.get("data", [{}])[0] if data.get("data") else {}
                return {
                    "success": True,
                    "quota_usage": quota_data.get("quota_usage"),
                    "quota_total": quota_data.get("config", {}).get("quota_total"),
                }
        except Exception as e:
            logger.error(f"[Instagram] get_publishing_limit error: {e}")
            return {"success": False, "error": str(e)}

File: lib/instagram.ts
/**
 * Instagram Graph API Service
 * https://developers.facebook.com/docs/instagram-api/guides/content-publishing
 *
 * Instagram Reels Publishing Requirements:
 * - Business or Creator Instagram account
 * - Connected Facebook Page
 * - 2-step process: Create container → Publish
 * - Video: 3-90 seconds, recommended ≤60 seconds
 * - Aspect ratio: 9:16 (vertical)
 * - Max file size: 1GB
 */

const INSTAGRAM_OAUTH_BASE = "https://api.instagram.com";
const INSTAGRAM_GRAPH_BASE = "https://graph.instagram.com";
const FACEBOOK_GRAPH_BASE = "https://graph.facebook.com/v21.0";

// Instagram API Types
export interface InstagramTokenResult {
  success: boolean;
  accessToken?: string;
  userId?: string;
  expiresIn?: number;
  error?: string;
  errorDescription?: string;
}

export interface InstagramUserInfo {
  id: string;
  username: string;
  name?: string;
  accountType: "BUSINESS" | "MEDIA_CREATOR" | "PERSONAL";
  profilePictureUrl?: string;
  followersCount?: number;
  mediaCount?: number;
}

export interface InstagramContainerResult {
  success: boolean;
  containerId?: string;
  error?: string;
  errorCode?: number;
  errorSubcode?: number;
}

export interface InstagramContainerStatus {
  status: "IN_PROGRESS" | "FINISHED" | "ERROR" | "EXPIRED" | "PUBLISHED";
  statusCode?: string;
}

export interface InstagramPublishResult {
  success: boolean;
  mediaId?: string;
  permalink?: string;
  error?: string;
  errorCode?: number;
}

export interface InstagramReelSettings {
  shareToFeed?: boolean;
  coverUrl?: string;
  thumbOffset?: number; // Cover frame offset in ms
  locationId?: string;
  collaboratorUsernames?: string[];
}

/**
 * Generate Instagram OAuth authorization URL
 * Uses Facebook Login for Instagram
 */
export function getInstagramAuthorizationUrl(
  appId: string,
  redirectUri: string,
  state: string
): string {
  const scopes = [
    "instagram_basic",
    "instagram_content_publish",
    "pages_show_list",
  ].join(",");

  const params = new URLSearchParams({
    client_id: appId,
    redirect_uri: redirectUri,
    scope: scopes,
    response_type: "code",
    state: state,
  });

  // Use Facebook OAuth for Instagram Business accounts
  return `https://www.facebook.com/v21.0/dialog/oauth?${params.toString()}`;
}

/**
 * Exchange authorization code for short-lived access token
 */
export async function exchangeInstagramCodeForToken(
  appId: string,
  appSecret: string,
  code: string,
  redirectUri: string
): Promise<InstagramTokenResult> {
  try {
    console.log("[Instagram] Exchanging code for token...");

    // Facebook OAuth token exchange
    const params = new URLSearchParams({
      client_id: appId,
      client_secret: appSecret,
      redirect_uri: redirectUri,
      code: code,
    });

    const response = await fetch(
      `${FACEBOOK_GRAPH_BASE}/oauth/access_token?${params.toString()}`
    );

    const data = await response.json();

    if (data.error) {
      console.error("[Instagram] Token exchange error:", data.error);
      return {
        success: false,
        error: data.error.message,
        errorDescription: data.error.error_user_msg,
      };
    }

    console.log("[Instagram] Got short-lived token, exchanging for long-lived...");

    // Exchange for long-lived token
    const longLivedResult = await exchangeForLongLivedToken(
      appId,
      appSecret,
      data.access_token
    );

    return longLivedResult;
  } catch (error) {
    console.error("[Instagram] exchangeInstagramCodeForToken error:", error);
    return {
      success: false,
      error: error instanceof Error ? error.message : "Unknown error",
    };
  }
}

/**
 * Exchange short-lived token for long-lived token (60 days)
 */
export async function exchangeForLongLivedToken(
  appId: string,
  appSecret: string,
  shortLivedToken: string
): Promise<InstagramTokenResult> {
  try {
    const params = new URLSearchParams({
      grant_type: "fb_exchange_token",
      client_id: appId,
      client_secret: appSecret,
      fb_exchange_token: shortLivedToken,
    });

    const response = await fetch(
      `${FACEBOOK_GRAPH_BASE}/oauth/access_token?${params.toString()}`
    );

    const data = await response.json();

    if (data.error) {
      console.error("[Instagram] Long-lived token exchange error:", data.error);
      return {
        success: false,
        error: data.error.message,
      };
    }

    return {
      success: true,
      accessToken: data.access_token,
      expiresIn: data.expires_in, // Usually 5184000 (60 days)
    };
  } catch (error) {
    console.error("[Instagram] exchangeForLongLivedToken error:", error);
    return {
      success: false,
      error: error instanceof Error ? error.message : "Unknown error",
    };
  }
}

/**
 * Refresh long-lived token (extends by 60 days)
 * Can only be done once per 24 hours
 */
export async function refreshInstagramToken(
  accessToken: string
): Promise<InstagramTokenResult> {
  try {
    const params = new URLSearchParams({
      grant_type: "ig_refresh_token",
      access_token: accessToken,
    });

    const response = await fetch(
      `${INSTAGRAM_GRAPH_BASE}/refresh_access_token?${params.toString()}`
    );

    const data = await response.json();

    if (data.error) {
      console.error("[Instagram] Token refresh error:", data.error);
      return {
        success: false,
        error: data.error.message,
      };
    }

    return {
      success: true,
      accessToken: data.access_token,
      expiresIn: data.expires_in,
    };
  } catch (error) {
    console.error("[Instagram] refreshInstagramToken error:", error);
    return {
      success: false,
      error: error instanceof Error ? error.message : "Unknown error",
    };
  }
}

/**
 * Get Instagram Business Account ID from Facebook Page
 */
export async function getInstagramAccountFromPage(
  accessToken: string
): Promise<{ success: boolean; instagramAccountId?: string; pageId?: string; error?: string }> {
  try {
    // First get user's pages
    const pagesResponse = await fetch(
      `${FACEBOOK_GRAPH_BASE}/me/accounts?access_token=${accessToken}`
    );
    const pagesData = await pagesResponse.json();

    if (pagesData.error) {
      return {
        success: false,
        error: pagesData.error.message,
      };
    }

    if (!pagesData.data || pagesData.data.length === 0) {
      return {
        success: false,
        error: "No Facebook Pages found. Instagram Business accounts require a connected Facebook Page.",
      };
    }

    // Get Instagram Business Account from first page with Instagram connected
    for (const page of pagesData.data) {
      const igResponse = await fetch(
        `${FACEBOOK_GRAPH_BASE}/${page.id}?fields=instagram_business_account&access_token=${accessToken}`
      );
      const igData = await igResponse.json();

      if (igData.instagram_business_account) {
        return {
          success: true,
          instagramAccountId: igData.instagram_business_account.id,
          pageId: page.id,
        };
      }
    }

    return {
      success: false,
      error: "No Instagram Business Account found connected to your Facebook Pages.",
    };
  } catch (error) {
    console.error("[Instagram] getInstagramAccountFromPage error:", error);
    return {
      success: false,
      error: error instanceof Error ? error.message : "Unknown error",
    };
  }
}

/**
 * Get Instagram account info
 */
export async function getInstagramUserInfo(
  accessToken: string,
  instagramAccountId: string
): Promise<{ success: boolean; data?: InstagramUserInfo; error?: string }> {
  try {
    const fields = "id,username,name,account_type,profile_picture_url,followers_count,media_count";
    const response = await fetch(
      `${INSTAGRAM_GRAPH_BASE}/${instagramAccountId}?fields=${fields}&access_token=${accessToken}`
    );

    const data = await response.json();

    if (data.error) {
      return {
        success: false,
        error: data.error.message,
      };
    }

    return {
      success: true,
      data: {
        id: data.id,
        username: data.username,
        name: data.name,
        accountType: data.account_type,
        profilePictureUrl: data.profile_picture_url,
        followersCount: data.followers_count,
        mediaCount: data.media_count,
      },
    };
  } catch (error) {
    console.error("[Instagram] getInstagramUserInfo error:", error);
    return {
      success: false,
      error: error instanceof Error ? error.message : "Unknown error",
    };
  }
}

/**
 * Step 1: Create a media container for Reels
 * Returns a container ID that needs to be polled for status
 */
export async function createReelsContainer(
  accessToken: string,
  instagramAccountId: string,
  videoUrl: string,
  caption: string,
  settings?: InstagramReelSettings
): Promise<InstagramContainerResult> {
  try {
    console.log("[Instagram] Creating Reels container...");

    const params: Record<string, string> = {
      media_type: "REELS",
      video_url: videoUrl,
      caption: caption.slice(0, 2200), // Instagram caption limit
      access_token: accessToken,
    };

    // Optional settings
    if (settings?.shareToFeed !== undefined) {
      params.share_to_feed = settings.shareToFeed.toString();
    }
    if (settings?.coverUrl) {
      params.cover_url = settings.coverUrl;
    }
    if (settings?.thumbOffset !== undefined) {
      params.thumb_offset = settings.thumbOffset.toString();
    }
    if (settings?.locationId) {
      params.location_id = settings.locationId;
    }
    if (settings?.collaboratorUsernames && settings.collaboratorUsernames.length > 0) {
      params.collaborators = settings.collaboratorUsernames.join(",");
    }

    const searchParams = new URLSearchParams(params);

    const response = await fetch(
      `${INSTAGRAM_GRAPH_BASE}/${instagramAccountId}/media?${searchParams.toString()}`,
      { method: "POST" }
    );

    const data = await response.json();

    if (data.error) {
      console.error("[Instagram] Create container error:", data.error);
      return {
        success: false,
        error: data.error.message,
        errorCode: data.error.code,
        errorSubcode: data.error.error_subcode,
      };
    }

    console.log("[Instagram] Container created:", data.id);
    return {
      success: true,
      containerId: data.id,
    };
  } catch (error) {
    console.error("[Instagram] createReelsContainer error:", error);
    return {
      success: false,
      error: error instanceof Error ? error.message : "Unknown error",
    };
  }
}

/**
 * Check container status
 * Poll until status is FINISHED or ERROR
 */
export async function checkContainerStatus(
  accessToken: string,
  containerId: string
): Promise<{ success: boolean; status?: InstagramContainerStatus; error?: string }> {
  try {
    const response = await fetch(
      `${INSTAGRAM_GRAPH_BASE}/${containerId}?fields=status,status_code&access_token=${accessToken}`
    );

    const data = await response.json();

    if (data.error) {
      return {
        success: false,
        error: data.error.message,
      };
    }

    return {
      success: true,
      status: {
        status: data.status,
        statusCode: data.status_code,
      },
    };
  } catch (error) {
    console.error("[Instagram] checkContainerStatus error:", error);
    return {
      success: false,
      error: error instanceof Error ? error.message : "Unknown error",
    };
  }
}

/**
 * Wait for container to be ready for publishing
 */
export async function waitForContainerReady(
  accessToken: string,
  containerId: string,
  maxAttempts: number = 30,
  intervalMs: number = 10000
): Promise<{ success: boolean; ready: boolean; error?: string }> {
  console.log("[Instagram] Waiting for container to be ready...");

  for (let attempt = 0; attempt < maxAttempts; attempt++) {
    const result = await checkContainerStatus(accessToken, containerId);

    if (!result.success) {
      return { success: false, ready: false, error: result.error };
    }

    const status = result.status?.status;
    console.log(`[Instagram] Container status (attempt ${attempt + 1}/${maxAttempts}): ${status}`);

    if (status === "FINISHED") {
      return { success: true, ready: true };
    }

    if (status === "ERROR" || status === "EXPIRED") {
      return {
        success: false,
        ready: false,
        error: `Container status: ${status} - ${result.status?.statusCode}`,
      };
    }

    // Wait before next check
    await new Promise((resolve) => setTimeout(resolve, intervalMs));
  }

  return {
    success: false,
    ready: false,
    error: "Container processing timed out",
  };
}

/**
 * Step 2: Publish the media container
 */
export async function publishContainer(
  accessToken: string,
  instagramAccountId: string,
  containerId: string
): Promise<InstagramPublishResult> {
  try {
    console.log("[Instagram] Publishing container...");

    const params = new URLSearchParams({
      creation_id: containerId,
      access_token: accessToken,
    });

    const response = await fetch(
      `${INSTAGRAM_GRAPH_BASE}/${instagramAccountId}/media_publish?${params.toString()}`,
      { method: "POST" }
    );

    const data = await response.json();

    if (data.error) {
      console.error("[Instagram] Publish error:", data.error);
      return {
        success: false,
        error: data.error.message,
        errorCode: data.error.code,
      };
    }

    console.log("[Instagram] Published successfully:", data.id);

    // Get permalink
    const mediaResponse = await fetch(
      `${INSTAGRAM_GRAPH_BASE}/${data.id}?fields=permalink&access_token=${accessToken}`
    );
    const mediaData = await mediaResponse.json();

    return {
      success: true,
      mediaId: data.id,
      permalink: mediaData.permalink,
    };
  } catch (error) {
    console.error("[Instagram] publishContainer error:", error);
    return {
      success: false,
      error: error instanceof Error ? error.message : "Unknown error",
    };
  }
}

/**
 * Full Reels publishing flow
 * Creates container → Waits for ready → Publishes
 */
export async function publishInstagramReel(
  accessToken: string,
  instagramAccountId: string,
  videoUrl: string,
  caption: string,
  hashtags: string[],
  settings?: InstagramReelSettings
): Promise<InstagramPublishResult> {
  try {
    // Combine caption with hashtags
    const fullCaption = hashtags.length > 0
      ? `${caption}\n\n${hashtags.map((tag) => (tag.startsWith("#") ? tag : `#${tag}`)).join(" ")}`
      : caption;

    // Step 1: Create container
    const containerResult = await createReelsContainer(
      accessToken,
      instagramAccountId,
      videoUrl,
      fullCaption,
      settings
    );

    if (!containerResult.success || !containerResult.containerId) {
      return {
        success: false,
        error: containerResult.error || "Failed to create container",
        errorCode: containerResult.errorCode,
      };
    }

    // Step 2: Wait for container to be ready
    const readyResult = await waitForContainerReady(
      accessToken,
      containerResult.containerId
    );

    if (!readyResult.success || !readyResult.ready) {
      return {
        success: false,
        error: readyResult.error || "Container not ready for publishing",
      };
    }

    // Step 3: Publish
    const publishResult = await publishContainer(
      accessToken,
      instagramAccountId,
      containerResult.containerId
    );

    return publishResult;
  } catch (error) {
    console.error("[Instagram] publishInstagramReel error:", error);
    return {
      success: false,
      error: error instanceof Error ? error.message : "Unknown error",
    };
  }
}

/**
 * Get publishing quota/limits
 */
export async function getPublishingLimit(
  accessToken: string,
  instagramAccountId: string
): Promise<{ success: boolean; quotaUsage?: number; quotaTotal?: number; error?: string }> {
  try {
    const response = await fetch(
      `${INSTAGRAM_GRAPH_BASE}/${instagramAccountId}/content_publishing_limit?fields=quota_usage,config&access_token=${accessToken}`
    );

    const data = await response.json();

    if (data.error) {
      return {
        success: false,
        error: data.error.message,
      };
    }

    return {
      success: true,
      quotaUsage: data.data?.[0]?.quota_usage,
      quotaTotal: data.data?.[0]?.config?.quota_total,
    };
  } catch (error) {
    console.error("[Instagram] getPublishingLimit error:", error);
    return {
      success: false,
      error: error instanceof Error ? error.message : "Unknown error",
    };
  }
}
